iterative/resource_runner.go

```diff
@@ -210,7 +211,17 @@ func resourceRunnerCreate(ctx context.Context, d *schema.ResourceData, m interfa
 	var logError error
 	var logEvents string
 	err = resource.Retry(d.Timeout(schema.TimeoutCreate), func() *resource.RetryError {
-		logEvents, logError = resourceMachineLogs(ctx, d, m)
+		switch cloud := d.Get("cloud").(string); cloud {
+		case "kubernetes":
+			logEvents, logError = resourceMachineLogs(ctx, d, m)
+		default:
+			logEvents, logError = utils.RunCommand("journalctl --unit cml --no-pager",
+				2*time.Second,
+				net.JoinHostPort(d.Get("instance_ip").(string), "22"),
+				"ubuntu",
+				d.Get("ssh_private").(string))
+		}
+
```

Comment:
I'd prefer to keep a separate `resourceMachineLogs` function for each provider, even if some of them share the same code. The proposed changes are introducing two additional calls for non-SSH vendors, and it doesn't feel quite elegant, in my opinion.

Reply:
IMHO having duplicate code is not elegant at all

Reply:
And of course we are having the resource machine which can have intermediate business logic

Reply:
> The proposed changes are introducing two additional calls for non-SSH vendors

Which two calls?

Reply:
> _IMHO having duplicate code is not elegant at all_

Agreed! Nevertheless, in this case we're just duplicating a line of code for the sake of having the same methods on every provider. We aren't copying and pasting a copy of the `utils.RunCommand` function, but just a single line call. 🙃 

> _And of course we are having the resource machine which can have intermediate business logic_

Intermediate busniess logic is a pretty general term here. While 66% of our supported vendors — or 75% very soon — are going to rely on SSH for log gathering, that percentage will considerably decrease as soon as we start adding more container orchestrators. Should we count a specific log gathering mechanism as _intemediate business logic_ just because it's the same for a couple of providers?

Reply:
> _Which two calls?_

For orchestrators and other non-SSH providers, we would be calling the following two blocks sequentially, hitting two identical `switch` statements (hello code duplication) and making two calls: one to `resourceMachineLogs` and other to `provider.ResourceMachineLogs`

https://github.com/iterative/terraform-provider-iterative/blob/05aa3a385ea295a6d3707441a7fb6651f2347fc7/iterative/resource_runner.go#L214-L223
https://github.com/iterative/terraform-provider-iterative/blob/05aa3a385ea295a6d3707441a7fb6651f2347fc7/iterative/resource_machine.go#L238-L249